# ![.right](https://lh6.ggpht.com/wUrTIhpHPVqw_VPCdhbSiF5TXlBlLyRHdH1gsII_y5NkgYKzbbj7cC2l6AGoVq-JN0U=w100 "MSDS 7331 Data Mining - Project 1")


# Project 2: Classification

---



# Team Members

- Chris Woodard
- Claire Chu
- Nathan Mowat
- Bill Kerneckel

---


# Rubic

- [Data Preperation 1](#dp1) 

- [Data Preperation 2](#dp2)

- [Modeling and Evaluation 1](#me1)

- [Modeling and Evaluation 2](#me2)

- [Modeling and Evaluation 3](#me3)

- [Modeling and Evaluation 4](#me4)

- [Modeling and Evaluation 5](#me5)

- [Modeling and Evaluation 6](#me6)

- [Deployment](#d)

- [Exceptional Work](#ew)

- [Appendix](#a)

<hr>



# OBJECTIVE: 
Using the NASA human resources dataset, let's use the variables to predict future records:
- what kind of model can we use to predict a "junior salary range level" classification, given the variables we have
- what kind of model can we use to predict a "caucasian" race classification, given the variables we have

# Data Preperation 1

---

10 Points - Define and prepare your class variables. Use proper variable representations (int, float, one-hot, etc.). Use pre-processing methods (as needed) for dimensionality reduction, scaling, etc. Remove variables that are not needed/useful for the analysis.

In [1]:
#load python libaries
import numpy as np
import pandas as pd
from pandas.tools.plotting import scatter_matrix
import matplotlib.pyplot as pl
import seaborn as sns
import glob
import sys
import os
import warnings
import ggplot
import datetime
import time
from datetime import date

In [2]:
%matplotlib inline

In [3]:
#load data file
data_file = pd.read_excel('/Users/wkerneck/Desktop/NEW_NASA_2006.xlsx') 
data_file2 = pd.read_excel('/Users/wkerneck/Desktop/NEW_NASA_2007.xlsx') 
data_file3 = pd.read_excel('/Users/wkerneck/Desktop/NASA_2008.xlsx')
data_file4 = pd.read_excel('/Users/wkerneck/Desktop/NASA_2009.xlsx')
data_file5 = pd.read_excel('/Users/wkerneck/Desktop/NASA_2010.xlsx')
data_file6 = pd.read_excel('/Users/wkerneck/Desktop/NASA_2011.xlsx')
data_file7 = pd.read_excel('/Users/wkerneck/Desktop/NASA_2012.xlsx')
data_file8 = pd.read_excel('/Users/wkerneck/Desktop/NASA_2013.xlsx')
data_file9 = pd.read_excel('/Users/wkerneck/Desktop/NASA_2014.xlsx')
data_file10 = pd.read_excel('/Users/wkerneck/Desktop/NASA_2015.xlsx')
data_file11 = pd.read_excel('/Users/wkerneck/Desktop/NASA_2016.xlsx')

#you'll have to change the file path to your working directory

#removing 1st row from datasets
data_file.drop(0, axis = 0,inplace = True)
data_file2.drop(0, axis = 0,inplace = True)
data_file3.drop(0, axis = 0,inplace = True)
data_file4.drop(0, axis = 0,inplace = True)
data_file5.drop(0, axis = 0,inplace = True)
data_file6.drop(0, axis = 0,inplace = True)
data_file7.drop(0, axis = 0,inplace = True)
data_file8.drop(0, axis = 0,inplace = True)
data_file9.drop(0, axis = 0,inplace = True)
data_file10.drop(0, axis = 0,inplace = True)
data_file11.drop(0, axis = 0,inplace = True)

#note: please be patient as this will take a few minutes to run

In [4]:
#combining dataframes 
df = [data_file, data_file2, data_file3, data_file4, data_file5, data_file6, data_file7, data_file8, data_file9, data_file10, data_file11]

cdf = pd.concat(df, axis=0, join='outer', join_axes=None, ignore_index=False,
          keys=None, levels=None, names=None, verify_integrity=False,
          copy=True)
cdf.reset_index()

index       Hist_yr Unnamed: 28   agency  bdyr coopsch coopyr  \
0           1          2006         NaN     NN51  1900     NaN    NaN   
1           2          2006         NaN     NN51  1980     NaN    NaN   
2           3          2006         NaN     NN23  1986     NaN    NaN   
3           4          2006         NaN     NN64  1994     NaN    NaN   
4           5          2006         NaN     NN72  2006     NaN    NaN   
5           6          2006         NaN     NN62  1972     NaN    NaN   
6           7          2006         NaN     NN51  1900     NaN    NaN   
7           8          2006         NaN     NN72  1986     NaN    NaN   
8           9          2006         NaN     NN51  1978     NaN    NaN   
9          10          2006         NaN     NN10  1987     NaN    NaN   
10         11          2006         NaN     NN62  1982     NaN    NaN   
11         12          2006         NaN     NN72  1976     NaN    NaN   
12         13          2006         NaN     NN22  1973     NaN    NaN   
13         14          2006         NaN     NN10  1975     NaN    NaN   
14         15          2006         NaN     NN76  1978     NaN    NaN   
15         16          2006         NaN     NN21  1980     NaN    NaN   
16         17          2006         NaN     NN21  1979     NaN    NaN   
17         18          2006         NaN     NN51  2003     NaN    NaN   
18         19          2006         NaN     NN10  1978     NaN    NaN   
19         20          2006         NaN     NN22  1983     NaN    NaN   
20         21          2006         NaN     NN62  1999     NaN    NaN   
21         22          2006         NaN     NN76  1900     NaN    NaN   
22         23          2006         NaN     NN51  1987     NaN    NaN   
23         24          2006         NaN     NN76  1995     NaN    NaN   
24         25          2006         NaN     NN51  2002     NaN    NaN   
25         26          2006         NaN     NN72  1967     NaN    NaN   
26         27          2006         NaN     NN10  1972     NaN    NaN   
27         28          2006         NaN     NN76  1982     NaN    NaN   
28         29          2006         NaN     NN51  1900     NaN    NaN   
29         30          2006         NaN     NN72  1996     NaN    NaN   
...       ...           ...         ...      ...   ...     ...    ...   
202707  17498          2016         NaN     NN51  1992     NaN      0   
202708  17499          2016         NaN     NN51  1992     NaN      0   
202709  17500          2016         NaN     NN23  1991     NaN      0   
202710  17501          2016         NaN     NN10  2005     NaN      0   
202711  17502          2016         NaN     NN23  1900     NaN      0   
202712  17503          2016         NaN     NN23  2008     NaN      0   
202713  17504          2016         NaN     NN72  1987     NaN      0   
202714  17505          2016         NaN     NN51  1997     NaN      0   
202715  17506          2016         NaN     NN51  1900     NaN      0   
202716  17507          2016         NaN     NN51  1900     NaN      0   
202717  17508          2016         NaN     NN76  1997     NaN      0   
202718  17509          2016         NaN     NN24  1997     NaN      0   
202719  17510          2016         NaN     NN51  1983     NaN      0   
202720  17511          2016         NaN     NN51  1991     NaN      0   
202721  17512          2016         NaN     NN10  1987     NaN      0   
202722  17513          2016         NaN     NN21  1900     NaN      0   
202723  17514          2016         NaN     NN72  2013     NaN      0   
202724  17515          2016         NaN     NN51  1989     NaN      0   
202725  17516          2016         NaN     NN62  2010     NaN      0   
202726  17517          2016         NaN     NN51  1987     NaN      0   
202727  17518          2016         NaN     NN51  2006     NaN      0   
202728  17519          2016         NaN     NN72  1990     NaN      0   
202729  17520          2016         NaN     NN21  1985     NaN    

In [5]:
# dropping columns we are not using for the analysis
cdf.drop('Unnamed: 28', axis=1, inplace=True)
cdf.drop('agency', axis=1, inplace=True)
cdf.drop('coopsch', axis=1, inplace=True)
cdf.drop('coopyr', axis=1, inplace=True)
cdf.drop('dtystn_ind', axis=1, inplace=True)
cdf.drop('dtystnname', axis=1, inplace=True)
cdf.drop('edlev', axis=1, inplace=True)
cdf.drop('foulev', axis=1, inplace=True)
cdf.drop('fousch', axis=1, inplace=True)
cdf.drop('fouyr', axis=1, inplace=True)
cdf.drop('nasattl', axis=1, inplace=True)
cdf.drop('nasa', axis=1, inplace=True)
cdf.drop('nasat', axis=1, inplace=True)
cdf.drop('probenddte', axis=1, inplace=True)
cdf.drop('promontedte', axis=1, inplace=True)
cdf.drop('secsch', axis=1, inplace=True)
cdf.drop('seclev', axis=1, inplace=True)
cdf.drop('typappt', axis=1, inplace=True)
cdf.drop('tl', axis=1, inplace=True)
cdf.drop('ttl', axis=1, inplace=True)
cdf.drop('worksch', axis=1, inplace=True)

In [6]:
cdf = cdf[cdf.bdyr != 1900]
cdf.salary = cdf.salary.astype(np.float64)
cdf.time_in_grade = cdf.time_in_grade.astype(np.float64)
cdf.suplev = cdf.suplev.astype(np.float64)

In [7]:
# converting bdyr column to age
#now = datetime.datetime.now()
now = date(2016,12,31)
cdf['bdyr'] = cdf['bdyr'].apply(pd.to_numeric, errors='coerce') 
cdf['age'] = now.year - cdf['bdyr']

In [8]:
cdf['salary_range'] = pd.cut(cdf['salary'],[0,50000,100000,135000,150000],4,labels=['Entry_Level','Jr_Level','Mid_Level', 'Mgmt_Level'])
print(cdf['salary_range'].head())
cdf.dropna(inplace=True)

1     Jr_Level
2    Mid_Level
3     Jr_Level
4    Mid_Level
5     Jr_Level
Name: salary_range, dtype: category
Categories (4, object): [Entry_Level < Jr_Level < Mid_Level < Mgmt_Level]


In [9]:
cdf.salary.max() #Is the salary range cut-off still valid with the new data? Yes

149993.0

In [10]:
# convert cdf variable back to "df" nomenclature
df = pd.DataFrame(cdf)

In [11]:
#manipulate employee start date 'eoddte' to get 'service' variable
#convert 'eoddte' to series
df.eod = pd.Series(df['eoddte'])
#convert 'eoddte' series to 'eoddte' datetime
df['eoddte'] = pd.to_datetime(df.eod) 
#convert eod to be just the year
df['eodyr'] = df['eoddte'].map(lambda x: x.year)

In [12]:
#convert high school graduation year "hiyr" to be numeric
df['hiyr'] = df['hiyr'].apply(pd.to_numeric, errors='coerce')
#after looking at a crosstab, we can see that there are a lot of "high school graduation year: 0"
#this is a data entry error, let's remove these entries...
df = df[df.hiyr != 0]
pd.crosstab(index=df["hiyr"],columns="hiyr")
#now the 0 entries have been removed

#subtract employee hire date from high school graduation date to get exprience
df['experience'] = df['eodyr'] - df['hiyr']

In [13]:
#get the 'service' variable by subtracting current date from hire date
today = date(2016, 12, 31)

df['service'] = today - df['eoddte']
#generate retirement potential by subtracting retirement eligibility date from todays date  
#convert retoptdte to series for conversion to datetime format
df.retoptdte = pd.Series(df['retoptdte'])
#convert 'retoptdte' series to 'eoddte' datetime
df['retoptdte'] = pd.to_datetime(df.retoptdte) 
#generate 'retpot' variable
df['retpot'] = today - df['retoptdte']

In [14]:
df_grouped = df.groupby(by=['grade','sex'])

In [15]:
df_grouped = df.groupby(by=['grade','sex'])
df_imputed = df_grouped.transform(lambda grp: grp.fillna(grp.median()))
df_imputed[['grade','sex']] = df[['grade','sex']]

In [16]:
df['Hist_yr'] = df['Hist_yr'].astype(np.float64)
df_yrgroup = df.groupby(by=['Hist_yr'])

In [17]:
# creating new dataframe. Keeping original seperate.
DF_NoT = df.copy()
# perform one-hot encoding of the categorical data "salary_range" and "rno" (ethnic background).
tmp_df = pd.get_dummies(df.salary_range,prefix='salary_range')
tmp_df2 = pd.get_dummies(df.rno,prefix='rno')

DF_NoT = pd.concat((df,tmp_df),axis=1) # add back into the dataframe
DF_NoT = pd.concat((DF_NoT,tmp_df2),axis=1)

In [18]:
# Delete time/dates - can't be used in L-Regression
del DF_NoT['eoddte']
del DF_NoT['retoptdte']
del DF_NoT['salary_range']
del DF_NoT['currgrddte']
del DF_NoT['frscdte']
del DF_NoT['fscdte']
del DF_NoT['hisch']
del DF_NoT['leqdte']
del DF_NoT['loccde']
del DF_NoT['ncc']
del DF_NoT['opmtitle']
del DF_NoT['occode']
del DF_NoT['orga']
del DF_NoT['orgabr']
del DF_NoT['orgadir']
del DF_NoT['orgadiv']
del DF_NoT['orgasec']
del DF_NoT['postenure']
del DF_NoT['tenure']
del DF_NoT['supind']
del DF_NoT['suplev']
del DF_NoT['rno']
del DF_NoT['rno_000010']  
del DF_NoT['rno_100100'] 
del DF_NoT['rno_100001'] 
del DF_NoT['rno_100010'] 
del DF_NoT['rno_100101'] 
del DF_NoT['rno_100000'] 
del DF_NoT['rno_011000'] 
del DF_NoT['rno_001101'] 
del DF_NoT['rno_010001'] 
del DF_NoT['rno_010101'] 
del DF_NoT['rno_001100'] 
del DF_NoT['rno_010000'] 
del DF_NoT['rno_010100'] 
del DF_NoT['rno_001001'] 
del DF_NoT['rno_001000'] 
del DF_NoT['rno_000101'] 
del DF_NoT['rno_000100'] 
del DF_NoT['rno_000011'] 
del DF_NoT['rno_101000'] 
del DF_NoT['rno_101101'] 
del DF_NoT['rno_110000'] 
del DF_NoT['rno_110001'] 
del DF_NoT['rno_110101'] 
del DF_NoT['rno_111111'] 

In [19]:
today = date(2016, 12, 31) #modify if more datasets are added
DF_NoT.service = DF_NoT.service / np.timedelta64(1, 'D')
DF_NoT.retpot = DF_NoT.retpot / np.timedelta64(1, 'D')



In [20]:
DF_NoT.lastpromodte = pd.Series(DF_NoT['lastpromodte'])
#convert 'eoddte' series to 'eoddte' datetime
DF_NoT['lastpromodte'] = pd.to_datetime(DF_NoT.lastpromodte) 
DF_NoT['lastpromodte'] = today - DF_NoT['lastpromodte']

#removing the day to calculate a number
DF_NoT.lastpromodte = DF_NoT.lastpromodte / np.timedelta64(1, 'D')

DF_NoT.nextwigdte = pd.Series(DF_NoT['nextwigdte'])
DF_NoT['nextwigdte'] = pd.to_datetime(DF_NoT.nextwigdte) 
DF_NoT['nextwigdte'] = today - DF_NoT['nextwigdte']
DF_NoT.nextwigdte = DF_NoT.nextwigdte / np.timedelta64(1, 'D')

DF_NoT.retdiscdte = pd.Series(DF_NoT['retdiscdte'])
DF_NoT['retdiscdte'] = pd.to_datetime(DF_NoT.retdiscdte) 
DF_NoT['retdiscdte'] = today - DF_NoT['retdiscdte']
DF_NoT.retdiscdte = DF_NoT.retdiscdte / np.timedelta64(1, 'D')

DF_NoT['IsMale'] = DF_NoT.sex=='M' 
DF_NoT.IsMale = DF_NoT.IsMale.astype(np.int)
del DF_NoT['sex']

# Data Preperation 2
---


5 Points - Describe the final dataset that is used for classification/regression (include a description of any newly formed variables you created).

In [43]:
DF_NoT.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 126967 entries, 2 to 17523
Data columns (total 49 columns):
Hist_yr                     126967 non-null float64
bdyr                        126967 non-null float64
grade                       126967 non-null object
hilev                       126967 non-null object
hiyr                        126967 non-null int64
install                     126967 non-null object
lastpromodte                126967 non-null float64
nextwigdte                  126967 non-null float64
retdiscdte                  126967 non-null float64
salary                      126967 non-null float64
secyr                       126967 non-null object
step_emp                    126967 non-null object
time_in_grade               126967 non-null float64
age                         126967 non-null float64
eodyr                       126967 non-null int64
experience                  126967 non-null int64
service                     126967 non-null float64
retpot           

Based on the output above, you can see we have about 126,967 entries with 41 columns. 
New Variables created are:
- the age category derived from the "birth year" variable
- the salary_range category derived by dividing up the salary variable into 4 categories
- the eodyr variable which shows the year of employment
- the experience variable derived from the high school graduation year subtracted from the last date of the data set (12/31/2016)
- the serivce variable derived from the year of employment subtracted from 12/31/2016
- the retirement potential derived from the date of earliest reitrement subtracted from 12/31/2016

Let's use the variables to predict future records in the NASA human resources data set
- which model can we use to predict a "junior salary range level" classification, given the variables we have
- which model can we use to predict a "caucasian" race classification, given the variables we have


# Modeling and Evaluation 1 - KNN, Decision Trees
---

10 Points - Choose and explain your evaluation metrics that you will use (i.e., accuracy,
precision, recall, F-measure, or any metric we have discussed). Why are the measure(s) appropriate for analyzing the results of your modeling? Give a detailed explanation backing up any assertions.

In [77]:

from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import StratifiedKFold
import numpy as np
from sklearn import metrics as mt

# X = create a matrix with all variable except response variable in it. rno_000001
# Y = this should be predictor variable
# create variables we are more familiar with
X = DF2_NoT.as_matrix()
target = len(DF2_NoT.salary_range_Jr_Level)

y = target
yhat = np.zeros(y.shape) # we will fill this with predictions

# create cross validation iterator
cv = StratifiedKFold(n_splits=10)

# get a handle to the classifier object, which defines the type
clf = KNeighborsClassifier(n_neighbors=3)

# now iterate through and get predictions, saved to the correct row in yhat
for train, test in cv.split(X,y):
    clf.fit(X[train],y[train])
    yhat[test] = clf.predict(X[test])

total_accuracy = mt.accuracy_score(y, yhat)
print ('KNN accuracy', total_accuracy)

AttributeError: 'int' object has no attribute 'shape'

# Modeling and Evaluation 2
---

10 Points - Choose the method you will use for dividing your data into training and testing splits (i.e., are you using Stratified 10-fold cross validation? Why?). Explain why your chosen method is appropriate or use more than one method as appropriate. For example, if you are using time series data then you should be using continuous training and testing sets across time

# Modeling and Evaluation 3
---

20 Points - Create three different classification/regression models for each task (e.g., random forest, KNN, and SVM for task one and the same or different algorithms for task two). Two modeling techniques must be new (but the third could be SVM or logistic regression). Adjust parameters as appropriate to increase generalization performance using your chosen metric. You must investigate different parameters of the algorithms!

# Modeling and Evaluation 4
---

10 Points - Analyze the results using your chosen method of evaluation. Use visualizations of the results to bolster the analysis. Explain any visuals and analyze why they are interesting to someone that might use this model.

# Modeling and Evaluation 5
---

10 Points - Discuss the advantages of each model for each classification task, if any. If there are not advantages, explain why. Is any model better than another? Is the difference significant with 95% confidence? Use proper statistical comparison methods. You must use statistical comparison techniques—be sure they are appropriate for your chosen method of validation as discussed in unit 7 of the course.

# Modeling and Evaluation 6
---

10 Points - Which attributes from your analysis are most important? Use proper methods discussed in class to evaluate the importance of different attributes. Discuss the results and hypothesize about why certain attributes are more important than others for a given classification task.

# Deployment
---


5 Points - How useful is your model for interested parties (i.e., the companies or organizations that might want to use it for prediction)? How would you measure the model's value if it was used by these parties? How would your deploy your model for interested parties? What other data should be collected? How often would the model need to be updated, etc.? 

# Exceptional Work
---

10 Points - You have free reign to provide additional analyses. One idea: grid search parameters in a parallelized fashion and visualize the performances across attributes. Which parameters are most significant for making a good model for each classification algorithm?

# Appendix